Detect out argument in the operator schema

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #64053
* #63651
* #63540
* #63096
* __->__ #62755




After this change, out argument can be checked by calling is_out(). We also update backward computability tests to make sure it works with both default and our arguments in the schema. 

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256/)

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256)

tugsbayasgalan committed:
Detect out argument in the schema

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Detect out argument in the schema"


After this change, out argument can be checked by calling is_out()

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Detect out argument in the schema"


After this change, out argument can be checked by calling is_out()

[ghstack-poisoned]

iseeyuan reviewed (approved):
Looks good to me.

iseeyuan commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -34,6 +34,8 @@ struct Argument {
         kwarg_only_(kwarg_only),
         alias_info_(std::move(alias_info)) {
+    bool is_alias = alias_info_.has_value() && alias_info_.value().isWrite();
+    is_out_ = kwarg_only_ && is_alias;
```
@gmagogsfm , could you confirm if this logic is good for all out args?

iseeyuan commented:
@cccclai , you may consume is_out() here: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/mobile/function.cpp#L100

gmagogsfm commented on test/cpp/jit/test_misc.cpp:
```diff
@@ -520,6 +520,14 @@ TEST(SchemaParserTest, NestedArrays) {
                                               .getElementType()));
 }
 
+TEST(SchemaParserTest, OutVariant) {
```
Could you add more test cases including:
1) operator without any out argument
2) `is_write` argument but not kwonly
3) arguments that are not `out` are correctly distinguished by `is_out`

tugsbayasgalan replied:
will do thanks

tugsbayasgalan committed:
Update on "Detect out argument in the schema"


After this change, out argument can be checked by calling is_out()

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Detect out argument in the schema"


After this change, out argument can be checked by calling is_out()

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30415256).

raziel reviewed (commented):
Curious if you think we should update the schema BC test function?

Commit:
Update on "Detect out argument in the schema"


After this change, out argument can be checked by calling is_out()

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256)

[ghstack-poisoned]

Commit:
Update on "Detect out argument in the schema"





After this change, out argument can be checked by calling is_out()

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256/)

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256)

[ghstack-poisoned]

Commit:
Update on "Detect out argument in the schema"





After this change, out argument can be checked by calling is_out()

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256/)

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256)

[ghstack-poisoned]

Commit:
Update on "Detect out argument in the schema"





After this change, out argument can be checked by calling is_out()

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256/)

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256)

[ghstack-poisoned]

Commit:
Update on "Detect out argument in the schema"





After this change, out argument can be checked by calling is_out()

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256/)

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256)

[ghstack-poisoned]

Commit:
Update on "Detect out argument in the schema"





After this change, out argument can be checked by calling is_out()

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256/)

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256)

[ghstack-poisoned]

Commit:
Update on "Detect out argument in the schema"





After this change, out argument can be checked by calling is_out()

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256/)

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256)

[ghstack-poisoned]

Commit:
Update on "Detect out argument in the schema"





After this change, out argument can be checked by calling is_out()

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256/)

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Detect out argument in the schema"





After this change, out argument can be checked by calling is_out()

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256/)

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30415256).

Commit:
Update on "Detect out argument in the schema"





After this change, out argument can be checked by calling is_out()

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256/)

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Detect out argument in the schema"





After this change, out argument can be checked by calling is_out()

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256/)

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30415256).

Commit:
Update on "Detect out argument in the schema"





After this change, out argument can be checked by calling is_out()

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256/)

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Detect out argument in the schema"





After this change, out argument can be checked by calling is_out()

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256/)

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30415256).

tugsbayasgalan committed:
Update on "Detect out argument in the schema"





After this change, out argument can be checked by calling is_out()

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256/)

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30415256).

gmagogsfm commented on aten/src/ATen/core/function_schema_inl.h:
```diff
@@ -121,17 +121,40 @@ inline bool FunctionSchema::isBackwardCompatibleWith(
+  // we want to test both out and default args seperately
+
+  // find the start of out args in the old schema
+  int old_out_start_idx;
```
Future clean up:
It feels like that this loop has appeared often enough to warrant a utility method, what do you think?

tugsbayasgalan replied:
will di

gmagogsfm commented on aten/src/ATen/core/function_schema_inl.h:
```diff
@@ -121,17 +121,40 @@ inline bool FunctionSchema::isBackwardCompatibleWith(
+
+  // find the start of out args in the old schema
+  int old_out_start_idx;
+  for (old_out_start_idx = old.arguments().size() - 1; old_out_start_idx > -1;
```
```suggestion
  for (old_out_start_idx = old.arguments().size() - 1; old_out_start_idx >= 0;
```

tugsbayasgalan replied:
will do

gmagogsfm commented on aten/src/ATen/core/function_schema_inl.h:
```diff
@@ -121,17 +121,40 @@ inline bool FunctionSchema::isBackwardCompatibleWith(
 
-  // Validate that all new arguments provided a default value.
-  for (size_t i = old.arguments().size(); i < arguments().size(); ++i) {
+  // Validate that all new arguments provided a default value
```
```suggestion
  // Validate that all new arguments provided has a default value
```

tugsbayasgalan replied:
thanks will do

tugsbayasgalan committed:
Update on "Detect out argument in the schema"





After this change, out argument can be checked by calling is_out(). We also update backward computability tests to make sure it works with both default and our arguments in the schema. 

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256/)

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30415256).

tugsbayasgalan committed:
Update on "Detect out argument in the schema"





After this change, out argument can be checked by calling is_out(). We also update backward computability tests to make sure it works with both default and our arguments in the schema. 

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256/)

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30415256).

tugsbayasgalan committed:
Update on "Detect out argument in the schema"





After this change, out argument can be checked by calling is_out(). We also update backward computability tests to make sure it works with both default and our arguments in the schema. 

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256/)

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30415256).

tugsbayasgalan committed:
Update on "Detect out argument in the schema"





After this change, out argument can be checked by calling is_out(). We also update backward computability tests to make sure it works with both default and our arguments in the schema. 

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256/)

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30415256).

Commit:
Update on "Detect out argument in the schema"





After this change, out argument can be checked by calling is_out(). We also update backward computability tests to make sure it works with both default and our arguments in the schema. 

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256/)

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Detect out argument in the schema"





After this change, out argument can be checked by calling is_out(). We also update backward computability tests to make sure it works with both default and our arguments in the schema. 

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256/)

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30415256).

tugsbayasgalan committed:
Update on "Detect out argument in the schema"





After this change, out argument can be checked by calling is_out(). We also update backward computability tests to make sure it works with both default and our arguments in the schema. 

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256/)

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256)

[ghstack-poisoned]

Commit:
Update on "Detect out argument in the schema"





After this change, out argument can be checked by calling is_out(). We also update backward computability tests to make sure it works with both default and our arguments in the schema. 

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256/)

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Detect out argument in the schema"





After this change, out argument can be checked by calling is_out(). We also update backward computability tests to make sure it works with both default and our arguments in the schema. 

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256/)

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256)

[ghstack-poisoned]

Commit:
Update on "Detect out argument in the schema"





After this change, out argument can be checked by calling is_out(). We also update backward computability tests to make sure it works with both default and our arguments in the schema. 

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256/)

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256)

[ghstack-poisoned]

Commit:
Update on "Detect out argument in the schema"





After this change, out argument can be checked by calling is_out(). We also update backward computability tests to make sure it works with both default and our arguments in the schema. 

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256/)

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Detect out argument in the schema"





After this change, out argument can be checked by calling is_out(). We also update backward computability tests to make sure it works with both default and our arguments in the schema. 

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256/)

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256)

[ghstack-poisoned]
